contracts/external/nexus/WrappedCoveredPrincipalToken.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.0;
+
+import { ERC20PermitWithSupply, ERC20Permit, IERC20Permit } from "../../libraries/ERC20PermitWithSupply.sol";
+import { Authorizable } from "../../libraries/Authorizable.sol";
+import { ITranche } from "../../interfaces/ITranche.sol";
+import { SafeERC20, IERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+/// @author Element Finance
+/// @title WrappedCoveredPrincipalToken
+contract WrappedCoveredPrincipalToken is ERC20PermitWithSupply, Authorizable {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    // Address of the base/underlying token which is used to buy the yield bearing token from the wrapped position.
+    // Ex - Dai is used to buy the yvDai yield bearing token
+    address public immutable baseToken;
+
+    // Enumerable address list, It contains the list of allowed tranches that are covered by this contract
+    // Criteria to choose the tranches are -
+    // a). Tranche should have same underlying/base token (i.e ETH, BTC, USDC).
+    // b). Should have the similar risk profiles.
+    EnumerableSet.AddressSet private _allowedTranches;
```

Comment:
I think we should move from using an enumerable set of tranches to an enumerable set of wrapped positions and then use create2 hash linking to validate that the tranche factory deployed to them. This would substantially reduce our overhead and responsibility as we won't need to add each new Tranche, instead if the wrapped position is approved any Tranche deployed by anyone one after would be covered.